eclair-core/src/main/scala/fr/acinq/eclair/payment/relay/ChannelRelayer.scala

```diff
@@ -56,35 +57,41 @@ class ChannelRelayer(nodeParams: NodeParams, relayer: ActorRef, register: ActorR
-          relayer ! Relayer.ForwardAdd(add, previousFailures :+ addFailed)
-        case _ =>
-          val failure = translateError(addFailed)
-          val cmdFail = CMD_FAIL_HTLC(originHtlcId, Right(failure), commit = true)
-          Metrics.recordPaymentRelayFailed(Tags.FailureType(cmdFail), Tags.RelayType.Channel)
-          log.info(s"rejecting htlc #$originHtlcId from channelId=$originChannelId reason=${cmdFail.reason}")
-          PendingRelayDb.safeSend(register, nodeParams.db.pendingRelay, originChannelId, cmdFail)
+    case addFailed@RES_ADD_FAILED(CMD_ADD_HTLC(_, _, _, _, _, Origin.ChannelRelayedHot(Upstream.ChannelRelayedHot(add, _), _), _, previousFailures), _, _) =>
+      log.info(s"retrying htlc #${add.id} from channelId=${add.channelId}")
+      relayer ! Relayer.RelayForward(add, previousFailures :+ addFailed)
+
+    case RES_ADD_COMPLETED(o: Origin.ChannelRelayedHot, htlc, fulfill: HtlcResult.Fulfill) =>
+      val upstream = o.upstream
+      val cmd = CMD_FULFILL_HTLC(upstream.originHtlcId, fulfill.paymentPreimage, commit = true)
+      PendingRelayDb.safeSend(register, nodeParams.db.pendingRelay, upstream.originChannelId, cmd)
+      context.system.eventStream.publish(ChannelPaymentRelayed(upstream.amountIn, upstream.amountOut, htlc.paymentHash, upstream.originChannelId, htlc.channelId))
+
+    case RES_ADD_COMPLETED(o: Origin.ChannelRelayedHot, _, fail: HtlcResult.Fail) =>
+      val upstream = o.upstream
+      Metrics.recordPaymentRelayFailed(Tags.FailureType.Remote, Tags.RelayType.Channel)
+      val cmd = fail match {
+        case f: HtlcResult.RemoteFail => CMD_FAIL_HTLC(upstream.originHtlcId, Left(f.fail.reason), commit = true)
+        case f: HtlcResult.RemoteFailMalformed => CMD_FAIL_MALFORMED_HTLC(upstream.originHtlcId, f.fail.onionHash, f.fail.failureCode, commit = true)
+        case _: HtlcResult.OnChainFail => CMD_FAIL_HTLC(upstream.originHtlcId, Right(PermanentChannelFailure), commit = true)
+        case f: HtlcResult.Disconnected => CMD_FAIL_HTLC(upstream.originHtlcId, Right(TemporaryChannelFailure(f.channelUpdate)), commit = true)
       }
-      case _ => throw new IllegalArgumentException(s"channel relayer received unexpected failure: $addFailed")
-    }
+      PendingRelayDb.safeSend(register, nodeParams.db.pendingRelay, upstream.originChannelId, cmd)
 
-    case ChannelCommandResponse.Ok => // ignoring responses from channels
+    case _: RES_SUCCESS[_] => // ignoring responses from channels
   }
 
   override def mdc(currentMessage: Any): MDC = {
     val paymentHash_opt = currentMessage match {
       case relay: RelayHtlc => Some(relay.r.add.paymentHash)
       case Register.ForwardShortIdFailure(Register.ForwardShortId(_, _, c: CMD_ADD_HTLC)) => Some(c.paymentHash)
-      case Status.Failure(addFailed: AddHtlcFailed) => Some(addFailed.paymentHash)
+      case addFailed: RES_ADD_FAILED[_] @unchecked => Some(addFailed.c.paymentHash)
```

Comment:
I don't think you need the `@unchecked`, do you?

Reply:
Done with c46fff704.